src/pages/index.js

```diff
@@ -17,47 +42,111 @@ popupImage.setEventListeners();
 const openPopupImage = (data) => {
   popupImage.open(data);
 };
-const renderCard = (card) => {
-  cardRenderer.addItem(createCard(card));
+
+const deletePopup = new PopupWithForm(constants.deletePopup, (data) => {
+  api
+    .deleteCard(data._id, data)
+    .then(() => {
+      deletePopup.submitButton.textContent = "Delete Card";
+      cardArray = document.querySelectorAll(".card");
+      cardArray.forEach((card) => {
+        if (card.id === data._id) {
+          card.remove();
+        }
+      });
+      deletePopup.close();
+    })
+    .catch((error) => {
+      alert(`Request to server failed. ${error}`);
+    });
+});
+deletePopup.setEventListeners();
+
+const openDeletePopup = (data) => {
+  deletePopup.open();
+  constants.deleteID.value = data._id;
 };
-const cardRenderer = new Section(
-  { items: constants.initialCards, renderer: renderCard },
-  constants.cardList
-);
 
-cardRenderer.renderItems();
+function sendLike(data) {
+  if (data.isLiked) {
+    api
+      .removeLike(data._id, data)
+      .then((data.isLiked = false))
+      .then(alert(`Like for "${data.name}" removed`));
+  } else {
+    api
+      .likeCard(data._id, data)
+      .then((data.isLiked = true))
+      .then(alert(`Like for "${data.name}" added`));
+  }
+}
 
 const userInfo = new UserInfo({
   name: constants.profileName,
-  description: constants.profileDesc,
+  about: constants.profileDesc,
 });
 const cardModal = new PopupWithForm(constants.cardModal, (data) => {
-  cardRenderer.addItem(createCard(data));
+  api
+    .sendCard(data)
+
+    .then((result) => {
+      cardRenderer.addItem(createCard(result));
+      cardModal.submitButton.textContent = "Create";
+      cardModal.close();
+    });
 });
+
 cardModal.setEventListeners();
+const avatarModal = new PopupWithForm(constants.changeAvatar, (data) => {
+  api.updateProfilePicture(data).then((result) => {
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```